modules/iam-eks-role/main.tf

```diff
@@ -1,58 +1,22 @@
-data "aws_caller_identity" "current" {}
-
-data "aws_partition" "current" {}
-
-data "aws_eks_cluster" "main" {
-  for_each = var.cluster_service_accounts
-
-  name = each.key
-}
-
 data "aws_iam_policy_document" "assume_role_with_oidc" {
-  dynamic "statement" {
-    for_each = var.cluster_service_accounts
-
-    content {
-      effect = "Allow"
-
-      actions = ["sts:AssumeRoleWithWebIdentity"]
-
-      principals {
-        type = "Federated"
-
-        identifiers = [
-          "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:oidc-provider/${replace(data.aws_eks_cluster.main[statement.key].identity[0].oidc[0].issuer, "https://", "")}"
-        ]
-      }
-
-      condition {
-        test     = "StringEquals"
-        variable = "${replace(data.aws_eks_cluster.main[statement.key].identity[0].oidc[0].issuer, "https://", "")}:sub"
-        values   = [for s in statement.value : "system:serviceaccount:${s}"]
-      }
-    }
-  }
```

Comment:
unfortunately, this is the crux of the issue with this sub-module however. 
1. its not clear to users that the cluster name should be the key of the maps
2. it restricts users to creating a role in one region while roles are global - they will will have to figure out when to use `cluster_service_accounts` and when to use `provider_url_sa_pairs` when its not quite clear what the difference is on the surface
3. most importantly is the lifecycle issue. you cannot specify this role and a cluster and create at the same time. the cluster has to exist first

Reply:
imho its not perfectly clear that the key of the map has to match the exact cluster name for functional reasons and isnt just meant to be an abbreviation to differentiate in the context of the map

Reply:
To give some context, this change came about from 
- https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1825
- https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1841

And I agree with the sentiments of those issues - these seem like common patterns many folks are repeating over and over and over (creating an IRSA role, specifying a policy, etc. - for common addons like VPC CNI, cluster autoscaler, etc.)

At first I thought something like this was being requested https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1827 but after clarifying, its really just the IRSA role w/ policy for common addons. Something where folks can just simply create an IRSA role and flip a flag to provide canned permissions for the use case of that role. 

So with that I came here because of the `iam-assumable-role-with-oidc` module, but then saw and remembered you added something for EKS and thought it would make the most sense to add on to this role to allow users to opt in to various canned permissions for common addons. However, in testing that I came across the lifecycle issues and  through testing cycles just reduced it down to the version you see now. 

Thats the context bit.

Regarding a path forward, I do believe we have to fix the lifecycle issue. Unfortunately this is how a lot of people work, especially when testing/trying out modules, plus we are seeing a lot more use cases where users are creating and destroying resources quite frequently in a CI pipeline type setup. So the targetted apply of `terraform apply -target module.vpc -target module.eks` is not going to fly. It also won't fly when we start propagating the use of this module to other modules and documentation. Its just one of the core tenants people rely on with Terraform - that Terraform will manage the lifecycle dependencies and not them.

Outside of that, personally I don't see the benefit in having the data source embedded with the restrictions it provides over having users provide the details. They can use the data source on their end if they like. And this is the area where I say that the module has been designed with in conjunction with the EKS module. I have aligned the inputs of this module with the outputs of that module so its very clear what is expected. That said, it doesn't lock people into using the EKS module; it just shows the relationship since I suspect this will most likely be the most used case.

Reply:
All the context being given here is from the behemoth EKS module, used solely by EKS admins. Where the `iam-eks-role` module was written for regular users who have workloads on EKS who just want a simple IRSA role created.